src/type_conversion/hiwire.c

```diff
@@ -404,53 +408,45 @@ EM_JS(int, hiwire_get_byteLength, (JsRef idobj), {
   return jsobj['byteLength'];
 });
 
-EM_JS(void, hiwire_copy_to_ptr, (JsRef idobj, int ptr), {
+EM_JS(void, hiwire_copy_to_ptr, (JsRef idobj, void* ptr), {
   var jsobj = Module.hiwire.get_value(idobj);
   // clang-format off
   var buffer = (jsobj['buffer'] !== undefined) ? jsobj.buffer : jsobj;
   // clang-format on
   Module.HEAPU8.set(new Uint8Array(buffer), ptr);
 });
 
-EM_JS(int, hiwire_get_dtype, (JsRef idobj), {
-  var jsobj = Module.hiwire.get_value(idobj);
-  switch (jsobj.constructor.name) {
-    case 'Int8Array':
-      dtype = 1; // INT8_TYPE;
-      break;
-    case 'Uint8Array':
-      dtype = 2; // UINT8_TYPE;
-      break;
-    case 'Uint8ClampedArray':
-      dtype = 3; // UINT8CLAMPED_TYPE;
-      break;
-    case 'Int16Array':
-      dtype = 4; // INT16_TYPE;
-      break;
-    case 'Uint16Array':
-      dtype = 5; // UINT16_TYPE;
-      break;
-    case 'Int32Array':
-      dtype = 6; // INT32_TYPE;
-      break;
-    case 'Uint32Array':
-      dtype = 7; // UINT32_TYPE;
-      break;
-    case 'Float32Array':
-      dtype = 8; // FLOAT32_TYPE;
-      break;
-    case 'Float64Array':
-      dtype = 9; // FLOAT64_TYPE;
-      break;
-    case 'ArrayBuffer':
-      dtype = 3;
-      break;
-    default:
-      dtype = 3; // UINT8CLAMPED_TYPE;
-      break;
-  }
-  return dtype;
-});
+EM_JS(void,
+      hiwire_get_dtype,
+      (JsRef idobj, char** format_ptr, Py_ssize_t* size_ptr),
+      {
+        if (!Module.hiwire.dtype_map) {
+          let entries = Object.entries({
+            'Int8Array' : [ 'b', 1 ],
+            'Uint8Array' : [ 'B', 1 ],
+            'Uint8ClampedArray' : [ 'B', 1 ],
+            'Int16Array' : [ "h", 2 ],
+            'Uint16Array' : [ "H", 2 ],
+            'Int32Array' : [ "i", 4 ],
+            'Uint32Array' : [ "I", 4 ],
+            'Float32Array' : [ "f", 4 ],
+            'Float64Array' : [ "d", 8 ],
+            'ArrayBuffer' : [ 'B', 1 ], // Default to Uint8;
+          });
+          Module.hiwire.dtype_map = new Map();
+          for (let[key, [ format, size ]] of entries) {
+            let format_utf8 =
+              allocate(intArrayFromString(format), "i8", ALLOC_NORMAL);
+            Module.hiwire.dtype_map.set(key, [ format_utf8, size ]);
+          }
```

Comment:
Arguably cleaner way of doing this:
```suggestion
          let alloc = c => allocate(intArrayFromString(c), "i8", ALLOC_NORMAL);
          Module.hiwire.dtype_map = new Map([
            ['Int8Array', [ alloc('b'), 1 ]],
            ['Uint8Array', [ alloc('B'), 1 ]],
            ['Uint8ClampedArray', [ alloc('B'), 1 ]],
            ['Int16Array', [ alloc('h'), 2 ]],
            ['Uint16Array', [ alloc('H'), 2 ]],
            ['Int32Array', [ alloc('i'), 4 ]],
            ['Uint32Array', [ alloc('I'), 4 ]],
            ['Float32Array', [ alloc('f'), 4 ]],
            ['Float64Array', [ alloc('d'), 8 ]],
            ['ArrayBuffer', [ alloc('B'), 1 ]], // Default to Uint8];
          ]);
```

Reply:
I found the API `stringToNewUTF8` which I think should be preferred over `allocate(intArrayFromString(string), "i8", ALLOC_NORMAL);`. Apparently there was some old chrome bug that meant we couldn't use this #141.

Reply:
I remember seeing that API but then failed to find it anywhere. Glad to know I was not crazy.